File: src/view/components/AbstractDisplayBlock.java
package view.components;

import model.structs.Rect;

public abstract class AbstractDisplayBlock {

    AbstractDisplayBlock parent;

    Rect margin;

    AlignType verticalAlign;

    AlignType horizontalAlign;

    public AbstractDisplayBlock(Rect margin, AlignType verticalAlign, AlignType horizontalAlign) {
        this.margin = margin;
        this.verticalAlign = verticalAlign;
        this.horizontalAlign = horizontalAlign;
    }

    public AbstractDisplayBlock() {
        this.margin = new Rect(0,0,0,0);
        this.verticalAlign = AlignType.MIDDLE;
        this.horizontalAlign = AlignType.BEGIN;
    }

    protected void setParent(AbstractDisplayBlock parent) {
        this.parent = parent;
    }

    protected void updateSizeRecursive(AbstractDisplayBlock child) {
        updateSize(child);
        if(parent != null) {
            parent.updateSizeRecursive(this);
        }
    }

    public Rect getMargin() {
        return margin;
    }

    public void setMargin(Rect margin) {
        this.margin = margin;
    }

    public void setVerticalAlign(AlignType verticalAlign) {
        this.verticalAlign = verticalAlign;
    }

    public void setHorizontalAlign(AlignType horizontalAlign) {
        this.horizontalAlign = horizontalAlign;
    }

    public AlignType getVerticalAlign() {
        return verticalAlign;
    }

    public AlignType getHorizontalAlign() {
        return horizontalAlign;
    }

    public int getOccupiedWidth() {
        return getAvailableWidth() + margin.left + margin.right;
    }

    public int getOccupiedHeight() {
        return getAvailableHeight() + margin.top + margin.bottom;
    }

    public abstract int getAvailableWidth();

    public abstract int getAvailableHeight();

    protected abstract void updateSize(AbstractDisplayBlock child);
}

File: src/view/components/AbstractDisplayBlock.kt
package view.components

import model.structs.Rect

abstract class AbstractDisplayBlock {
    var parent: AbstractDisplayBlock? = null

    var margin: Rect

    var verticalAlign: AlignType?

    var horizontalAlign: AlignType?

    constructor(margin: Rect, verticalAlign: AlignType?, horizontalAlign: AlignType?) {
        this.margin = margin
        this.verticalAlign = verticalAlign
        this.horizontalAlign = horizontalAlign
    }

    constructor() {
        this.margin = Rect(0, 0, 0, 0)
        this.verticalAlign = AlignType.MIDDLE
        this.horizontalAlign = AlignType.BEGIN
    }

    protected fun updateSizeRecursive(child: AbstractDisplayBlock?) {
        updateSize(child)
        if (parent != null) {
            parent!!.updateSizeRecursive(this)
        }
    }

    val occupiedWidth: Int
        get() = this.availableWidth + margin.left + margin.right

    val occupiedHeight: Int
        get() = this.availableHeight + margin.top + margin.bottom

    abstract val availableWidth: Int

    abstract val availableHeight: Int

    protected abstract fun updateSize(child: AbstractDisplayBlock?)
}
